go/vt/vtgate/planbuilder/select2.go

```diff
@@ -0,0 +1,200 @@
+	"errors"
+
+	"github.com/youtube/vitess/go/vt/sqlparser"
+)
+
+// PlanBuilder represents any object that's used to
+// build a plan.
+type PlanBuilder interface{}
+
+// JoinBuilder is used to build a Join primitive.
+type JoinBuilder struct {
+	IsLeft      bool
+	Left, Right PlanBuilder
+}
+
+// RouteBuilder is used to build a Route primitive.
+type RouteBuilder struct {
+	From     sqlparser.TableExpr
+	ChunkNum int
+	Route    *Route
+}
+
+// Route represents a Route primitive.
+type Route struct {
+	PlanID   PlanID
+	Keyspace *Keyspace
+	Vindex   Vindex      `json:",omitempty"`
+	Values   interface{} `json:",omitempty"`
+}
+
+// MarshalJSON marshals RouteBuilder into a readable form.
+func (rtb *RouteBuilder) MarshalJSON() ([]byte, error) {
+	marshalRoute := struct {
+		From     string `json:",omitempty"`
+		ChunkNum int    `json:",omitempty"`
+		Route    *Route
+	}{
+		From:     sqlparser.String(rtb.From),
+		ChunkNum: rtb.ChunkNum,
+		Route:    rtb.Route,
+	}
+	return json.Marshal(marshalRoute)
+}
+
+// buildSelectPlan2 is the new function to build a Select plan. It will be renamed
+// after the old one is removed.
+func buildSelectPlan2(sel *sqlparser.Select, schema *Schema) (PlanBuilder, *SymbolTable, error) {
+	return processTableExprs(sel.From, schema)
+}
+
+// processTableExprs is the root function to process the FROM clause. It produces a PlanBuilder
+// and the associated SymbolTable. The PlanBuilder is a tree of JoinBuilder and RouteBuilder
+// objects, where all the chunks are identified, and each chunk is grouped under one RouteBuilder.
+func processTableExprs(tableExprs sqlparser.TableExprs, schema *Schema) (PlanBuilder, *SymbolTable, error) {
+	if len(tableExprs) != 1 {
+		return nil, nil, errors.New("no list")
+	}
+	return processTableExpr(tableExprs[0], schema)
+}
+
+// processTableExpr produces a PlanBuilder subtree and SymbolTable for the given TableExpr.
+func processTableExpr(tableExpr sqlparser.TableExpr, schema *Schema) (PlanBuilder, *SymbolTable, error) {
+	switch tableExpr := tableExpr.(type) {
+	case *sqlparser.AliasedTableExpr:
+		return processAliasedTable(tableExpr, schema)
+	case *sqlparser.ParenTableExpr:
+		planBuilder, symbols, err := processTableExprs(tableExpr.Exprs, schema)
+		if route, ok := planBuilder.(*RouteBuilder); ok {
+			route.From = tableExpr
+		}
+		return planBuilder, symbols, err
+	case *sqlparser.JoinTableExpr:
+		return processJoin(tableExpr, schema)
+	}
+	panic("unreachable")
+}
+
+// processAliasedTable produces a PlanBuilder subtree and SymbolTable for the given AliasedTableExpr.
+func processAliasedTable(tableExpr *sqlparser.AliasedTableExpr, schema *Schema) (PlanBuilder, *SymbolTable, error) {
+	switch expr := tableExpr.Expr.(type) {
+	case *sqlparser.TableName:
+		route, table, err := getTablePlan(expr, schema)
+		if err != nil {
+			return nil, nil, err
+		}
+		planBuilder := &RouteBuilder{
+			From:  tableExpr,
+			Route: route,
+		}
+		alias := expr.Name
+		if tableExpr.As != "" {
```

Comment:
This is not the right place, though. I'll add it as comment where AliasedTableExpr is defined.